java-manta-client-unshaded/src/main/java/com/joyent/manta/client/crypto/ExternalSecurityProviderLoader.java

```diff
@@ -84,7 +124,90 @@ public static Provider getBouncyCastleProvider() {
         return BC_PROVIDER;
     }
 
+    public static Provider getSunJceProvider() {
+        return SUNJCE_PROVIDER;
+    }
+
+    /**
+     * Checks to see if a given provider as specified by name is a known and
+     * valid security provider.
+     *
+     * @param provider name/id of security provider
+     * @return true if known and valid, otherwise false
+     */
+    private static boolean isKnownSecurityProvider(final String provider) {
+        if (StringUtils.isBlank(provider)) {
+            return false;
+        }
+
+        switch (provider) {
+            case PKCS11_PROVIDER_NAME:
+            case BC_PROVIDER_NAME:
+            case SUNJCE_PROVIDER_NAME:
+                return Security.getProvider(provider) != null;
+            default:
+                return false;
+        }
+    }
+
+    /**
+     * Chooses the default security provider based on an internal ranking
+     * (pkcs11, Bouncy Castle, SunJCE) or the JVM system property
+     * <pre>manta.preferred.security.provider</pre>.
+     *
+     * @return the security provider to use for client side encryption
+     */
     public static Provider getPreferredProvider() {
-        return ObjectUtils.firstNonNull(PKCS11_PROVIDER, BC_PROVIDER);
+        return rankedPreferredProviders.get(0);
+    }
+
+    public static List<Provider> getRankedPreferredProviders() {
+        return rankedPreferredProviders;
+    }
+
+    public static void setRankedPreferredProviders(final Collection<Provider> providersRanked) {
+        rankedPreferredProviders = Collections.unmodifiableList(
+                new ArrayList<>(providersRanked));
+    }
+
+    /**
+     * @param preferredProvidersCSV CSV list of security provider names
+     * @return a list of valid security providers ranked from most preferred to least
+     */
+    public static List<Provider> buildRankedPreferredProviders(final String preferredProvidersCSV) {
+        final ArrayDeque<Provider> rankedProviders = new ArrayDeque<>(4);
+
+        if (PKCS11_PROVIDER != null) {
+            rankedProviders.add(PKCS11_PROVIDER);
+        }
+
+        if (BC_PROVIDER != null) {
+            rankedProviders.add(BC_PROVIDER);
+        }
+
+        if (SUNJCE_PROVIDER != null) {
+            rankedProviders.add(SUNJCE_PROVIDER);
+        }
+
+        final String[] providers;
+
+        if (StringUtils.isNotBlank(preferredProvidersCSV)) {
+            providers = StringUtils.split(preferredProvidersCSV, ",");
+        } else {
+            providers = new String[0];
+        }
+
+        ArrayUtils.reverse(providers);
+        Arrays.stream(providers)
+                .map(StringUtils::trimToEmpty)
+                .filter(StringUtils::isNotBlank)
+                .filter(ExternalSecurityProviderLoader::isKnownSecurityProvider)
```

Comment:
This filter will just drop unknown strings, right?  I'm not sure if silent ignoring or an error is better on invalid configuration.

Reply:
I've now added warnings to it. I'm not sure that we want to stop loading the library on this setting being fat fingered. Normally, I would be all for it, but system properties being misset in Java typically just have the JVM revert to its default behavior.